### ISSUE TITLE
Enclave threads should be terminated on exit

### ISSUE THREAD

**Issue by jethrogb** [fortanix/rust-sgx#165]
This program may not terminate:

```rust
use std::{sync::mpsc, thread};

fn main() {
    let (tx, rx) = mpsc::channel();
    thread::spawn(move || {
        tx.send(()).unwrap();
        loop {}
    });
    rx.recv().unwrap();
    println!("Exiting main thread");
}
```

To fix:

1. Change the AEP so that it won't automatically re-enter the enclave after AEX.
2. Add a way to return `EnclaveAbort` or such from `coenter` when another thread does the `exit` usercall.

**Comment by jethrogb** [comment]
> Change the AEP so that it won't automatically re-enter the enclave after AEX.
Hope this is possible with VDSO?

**Comment by raoulstrackx** [comment]
Unfortunately, the `user_handler` in the `SgxEnclaveRun` of the vdso isn't called upon each `enclu[eresume]`, nor when a signal is send to the thread. See this [poc](https://github.com/fortanix/rust-sgx/commits/raoul/gh-165-poc_vdso_usercall_handler).

**Comment by jethrogb** [comment]
That's unfortunate, I do remember some discussion about this on LKML but not the resolution.

Your POC seems unnecessarily complex. Why don't you just pass `sgx_user_handler` directly in `SgxEnclaveRun`?

Anyway, that means step 1 should be changed to use a signal handler instead of a different AEP.

**Comment by raoulstrackx** [comment]
> Your POC seems unnecessarily complex. Why don't you just pass `sgx_user_handler` directly in `SgxEnclaveRun`?
Yes that's the first thing I tried. However, he [vdso always calls](https://github.com/torvalds/linux/blob/dad9774deaf1cf8e8f7483310dfb2690310193d2/arch/x86/entry/vdso/vsgx.S#L75) the `sgx_user_handler` when the enclave exits. By the time it returns safe-by-caller registers may already have changed, but our code uses those registers to [communicate](https://github.com/fortanix/rust-sgx/compare/master...raoul/gh-165-poc_vdso_usercall_handler#diff-0443940e4ad302e1391f854019e99ee122efe7065f3f480b2a7957beda0e7c23L151-L157) with the enclave. The easiest option for this poc (without relying on compiler assumptions) was to wrap the caller in a code block that saves these values before calling the `sgx_user_handler`.

> Anyway, that means step 1 should be changed to use a signal handler instead of a different AEP.
Yes I've developed a [POC](https://github.com/fortanix/rust-sgx/commits/raoul/gh-165-signal_enclave_termination_vdso) for that as well

**Comment by mkaynov** [comment]
@raoulstrackx @jethrogb
I made an [implementation](https://github.com/fortanix/rust-sgx/compare/master...max/gh165_support_for_enclave_abort_non_vdso) that allows interrupting the hanged enclave thread by rewriting AEP, despite some open TODOs in the code this technique is proven to work, though only for non-vdso case.

For the VDSO none of these things are working :
- AEP rewriting: not possible
- Signal handler: got executed by the main thread regardless of the thread id signal being issued to
- custom sgx_enclave_user_handler_t: it is not triggered by AEX at all
Are there any other ideas in mind to make use of?
